All works, just need to satisfy mypy and whatnot now

charles-turner-1 opened the pull request:
- [ ] Closes #xxxx
- [x] Tests added

jemmajeffree commented:
TLDR: don't mind me, it's not going to cause any issues

Firstly, what I thought was a cftimedelta turned out to be a numpy timedelta hanging out with a cftime
<img width="654" height="256" alt="Screenshot 2025-07-14 at 5 23 31 pm" src="https://github.com/user-attachments/assets/c00c7653-2b72-4e5b-bd9e-e97b7a6a80fe" />
When I did manage to coerce this timedelta into cftime conventions, it just contained a floating point number of days, so I can't see anything having issues with its size
```python
coder = xr.coding.times.CFTimedeltaCoder()
result = coder.encode(oops.average_DT).load()
print(result.dtype)
result
```
<img width="544" height="167" alt="Screenshot 2025-07-14 at 5 38 33 pm" src="https://github.com/user-attachments/assets/44e716d7-2004-48b2-a697-dd708d378a2b" />

dcherian commented on xarray/structure/chunks.py:
```diff
@@ -83,9 +84,27 @@ def _get_chunk(var: Variable, chunks, chunkmanager: ChunkManagerEntrypoint):
+
+        from xarray.namedarray.utils import build_chunkspec
+
+        target_chunksize = parse_bytes(dask_config.get("array.chunk-size"))
```
How about adding `get_auto_chunk_size` to the ChunkManager class; and put the dask-specific stuff in the DaskManager.

cc @TomNicholas

dcherian commented on xarray/tests/test_backends.py:
```diff
@@ -5427,6 +5427,35 @@ def test_open_multi_dataset(self) -> None:
             ) as actual:
                 assert_identical(expected, actual)
 
+    def test_open_dataset_cftime_autochunk(self) -> None:
```
```suggestion
    @requires_cftime
    def test_open_dataset_cftime_autochunk(self) -> None:
```

dcherian commented on xarray/tests/test_backends.py:
```diff
@@ -5427,6 +5427,35 @@ def test_open_multi_dataset(self) -> None:
+            },
+            {"time": [cftime.Datetime360Day(2005, 12, 1, 12, 0, 0, 0)]},
+        )
+        with create_tmp_file() as tmp:
+            original.to_netcdf(tmp)
+            with open_dataset(tmp, chunks="auto") as actual:
```
```suggestion
        with self.roundtrip(original, open_kwargs={"chunks": "auto"}) as actual:
```

dcherian commented on xarray/namedarray/daskmanager.py:
```diff
@@ -264,3 +268,63 @@ def shuffle(
+
+        return data.rechunk(chunks, **kwargs)
+
+    def get_auto_chunk_size(self, var: Variable) -> tuple[int, _DType]:
+        from dask import config as dask_config
+        from dask.utils import parse_bytes
+
+        from xarray.namedarray.utils import fake_target_chunksize
+
+        target_chunksize = parse_bytes(dask_config.get("array.chunk-size"))
+        return fake_target_chunksize(var, target_chunksize=target_chunksize)
```
```suggestion
        return parse_bytes(dask_config.get("array.chunk-size"))
```

Only this is dask-specific, so that's what the DaskManager should be responsible for.

dcherian replied:
```suggestion
    def get_auto_chunk_size(self) -> int:
        from dask import config as dask_config
        from dask.utils import parse_bytes

        return parse_bytes(dask_config.get("array.chunk-size"))
```
Only this much is dask-specific, so that's what the DaskManager should be responsible for.

dcherian commented on xarray/structure/chunks.py:
```diff
@@ -83,8 +86,22 @@ def _get_chunk(var: Variable, chunks, chunkmanager: ChunkManagerEntrypoint):
+    auto_chunks = any(_chunk == "auto" for _chunk in _chunks)
+
+    limit = chunkmanager.get_auto_chunk_size()
+    no_op = not (_contains_cftime_datetimes(var) and auto_chunks)
+    limit, var_dtype = fake_target_chunksize(var, limit, no_op)
```
```suggestion
    if _contains_cftime_datetimes(var) and auto_chunks:
        limit, var_dtype = fake_target_chunksize(var, limit)
```

charles-turner-1 replied:
Done

dcherian commented on xarray/namedarray/utils.py:
```diff
@@ -195,6 +198,42 @@ def either_dict_or_kwargs(
+    if no_op:
+        return target_chunksize, data.dtype
+
+    import numpy as np
```
let's move imports to the top if we can; and remove the `no_op` bit

dcherian replied:
What do you mean?

charles-turner-1 replied:
Looked this up subsequently and I think I'm talking waffle -  the `if no_op` was just in there so that the logic for getting the array item size (in bytes) from the first item was skipped if we didn't find a cftime dtype in the array and a request for auto chunking.

Since arrays can only contain a single dtype, this shouldn't make any difference.

TLDR; ignore my previous comment, it was nonsense

dcherian commented on xarray/namedarray/utils.py:
```diff
@@ -195,6 +198,42 @@ def either_dict_or_kwargs(
+    the ratio of that size to a np.float64. Then we can just adjust our target_chunksize
+    and use the default dask chunking algorithm to get a reasonable chunk size.
+
+    ? I don't think T_chunkedArray or Variable should be necessary, but the calls
+    ? to this in daskmanager.py requires it to be that. I still need to wrap my head
+    ? around the typing here a bit more.
```
Can we delete this?

charles-turner-1 replied:
yup

dcherian commented on xarray/tests/test_dask.py:
```diff
@@ -1153,6 +1176,14 @@ def test_auto_chunk_da(obj):
     assert actual.chunks == expected.chunks
 
 
+@pytest.mark.parametrize("obj", [make_da_cftime()])
+def test_auto_chunk_da_cftime(obj):
```
Since this is only used in one place, let's just put the definition in the test.

charles-turner-1 replied:
Done

dcherian commented:
Sorry for the late review here. I left a few minor comments. Happy to merge after those are addressed

charles-turner-1 commented on xarray/tests/test_backends.py:
```diff
@@ -6060,6 +6060,34 @@ def test_open_multi_dataset(self) -> None:
+        ensure that auto-chunking works correctly."""
+        import cftime
+
+        from xarray.core.common import _contains_cftime_datetimes
```
done

charles-turner-1 commented:
No worries, figured you must have been busy/on holiday. 

I've addressed all those comments - thanks for all the help getting off the ground with this!

charles-turner-1 commented on xarray/structure/chunks.py:
```diff
@@ -83,8 +85,15 @@ def _get_chunk(var: Variable, chunks, chunkmanager: ChunkManagerEntrypoint):
     )
 
+    limit = chunkmanager.get_auto_chunk_size()
+    limit, var_dtype = fake_target_chunksize(var, limit)
```
I guess with the current implementation `maybe_fake_target_chunksize` would be a better name for the function, if we revert https://github.com/pydata/xarray/commit/b5933ed2d6e0b070b97b544d6c80e62c152a2ed4 then `fake_target_chunksize` makes more sense again.

dcherian replied:
I agree the name could be improved but don't have any good ideas at the moment.

(sorry for the delayed response here)

dcherian commented on xarray/namedarray/utils.py:
```diff
@@ -195,6 +198,31 @@ def either_dict_or_kwargs(
+
+    output_dtype = np.dtype(np.float64)
+
+    if data.dtype == object:
+        nbytes_approx: int = sys.getsizeof(first_n_items(data, 1))  # type: ignore[no-untyped-call]
+    else:
+        nbytes_approx = data.dtype.itemsize
```
```suggestion
        return limit, var.dtype
```

Otherwise we're doing the rescaling for all variables no?

dcherian replied:
```suggestion
    if data.dtype != object:
        return limit, var.dtype
        
    nbytes_approx: int = sys.getsizeof(first_n_items(data, 1))  # type: ignore[no-untyped-call]
```

charles-turner-1 commented:
It looks like the failing test (same one I commented on above) might be flaky? Now only faling for windows & python3.11, not 3.13: https://github.com/pydata/xarray/actions/runs/17962030991/job/51087173288

dcherian commented:
There's this one https://github.com/pydata/xarray/pull/9897 ;) but it's a bit gnarly, high impact though.

charles-turner-1 commented on xarray/namedarray/utils.py:
```diff
@@ -195,6 +200,106 @@ def either_dict_or_kwargs(
+
+    output_dtype = np.dtype(np.float64)
+
+    nbytes_approx: int = sys.getsizeof(first_n_items(data, 1))  # type: ignore[no-untyped-call]
```
This is for the assumed size as a float64 right? I think what you're saying is true but the array overhead rapidly becomes unimportant for reasonably large arrays? Very rough & ready analysis below:
```python
import matplotlib.pyplot as plt

sizes : list[tuple[int,int]] = []

for n in [1,10,100,1000,10000,100000,1000000,10000000]:
    arr = np.zeros(n, dtype=np.float64)
    sizes.append((n, sys.getsizeof(arr)/n))

# Plot size per element vs number of elements
plt.figure(figsize=(10,6))

plt.plot([n for n,_ in sizes], [size for _,size in sizes], marker='o')

plt.xscale('log')

plt.xlabel('Number of elements in array (log scale)')
plt.ylabel('Size per element (bytes)')

# Add 8 byte line
plt.axhline(y=8, color='r', linestyle='--', label='8')
```
<img width="850" height="529" alt="image" src="https://github.com/user-attachments/assets/f2105b8b-3af7-4398-b784-be2ade894145" />

jsignell replied:
Exactly! Calling `sys.getsizeof` on an array containing a single `cftime` object is not going to be a good representation of the memory consumption of an array of these things. Even if you pop the object out of the array that is still not really a good representation of the memory consumption. I think you'd do better with just `nbytes_approx: int = 8`

I made that same plot you did but with cftimes inside the array:

```python
# Does this still matter for decently sized arrays?
import sys
import numpy as np
import matplotlib.pyplot as plt

sizes : list[tuple[int,int]] = []

for n in np.logspace(0, 4, num=50, dtype=int):
    arr = np.array([cftime.DatetimeGregorian.fromordinal(2450000+i) for i in range(n)])
    sizes.append((n, sys.getsizeof(arr)/n))

# Plot size per element vs number of elements
plt.figure(figsize=(10,6))


plt.plot([n for n,_ in sizes], [size for _,size in sizes], marker='o')

plt.xscale('log')

plt.xlabel('Number of elements in array (log scale)')
plt.ylabel('Size per element (bytes)')

# Add 8 byte line
plt.axhline(y=8, color='r', linestyle='--', label='8')
```
<img width="860" height="531" alt="image" src="https://github.com/user-attachments/assets/8ed2385f-3547-4cfa-a639-97ba796359b4" />

charles-turner-1 replied:
My bad - I thought we'd popped the first cftime element out of the array and had a look at its size at that point.

It looks like the cftime elements in the array are 8bytes too - is that what we expect? I would have expected them to be a bit larger due to the extra overhead...

Assuming I'm wrong about that, it would be much simpler to just tell dask that a cftime is 8 bytes and leave the limit unadjusted - the ratio of two line should be pretty much 1 for all decently sized arrays. 

On my phone right now but I'll have a proper look when I get to my computer.
